### PR TITLE
Implement TLS certificate pinning and Fallback Address

### PR DESCRIPTION
**Pull request by lucebac**
This PR adds two features to Velociraptor:
* TLS Certificate Pinning: if one specifies the following snippet in the Client configuration, Velocirpator enables certificate pinning for both it's API connections and HTTP(S) connections established by vql.networking.GetHTTPTransport.
```
Client:
  certificate_hashes:
    - 10b5e383aec607420511a30ae520d500e81a50986e3f83a8fb407746060a916c
    - c1033930d035804c8cc74be7e797068fc00973344c9637869f27054088e75a87
  # only if the following value is set to true, strict cert pinning is enforced and 
  # connections are aborted if no fingerprints match; otherwise the fingerprints 
  # are only used to implement some kind of 'early-accept' for connections to 
  # bypass certificate validation for self-signed certificates
  force_certificate_pinning: true
```
* Fallback Addresses: we experienced several situations where we could not use Velicoraptor in customer's networks because of DNS being unavailable. This PR introduces the concept of fallback addresses which Velociraptor can fall back to when the connection to the primary address could not be established. The idea is roughly as follows: first, try to connecto to the original target; if that did not work, try the fallback address (if configured). Considering the example below, whenever a connection to `my-velociraptor-server.com:443` could not be established, `123.123.123.123:443` would be tried as a fallback.
```
Client:
  fallback_addresses:
    "my-velociraptor-server.com:443": "123.123.123.123:443"
```

Please note: both the certificate pinning and the fallback addresses currently only apply to connections which are established by `vql.networking.GetHttpTransport`. Other connections don't take the fallback addresses into account. I did intentionally not overwrite `net.http.DefaultTransport` as I am not able to tell whether or not this would break other parts of the code.

During the implementation of the TLS cert pinning, I experienced that the API connections used a different TLS config and HTTP Transport than the VQL. I took the chance to align both code paths so that the connection settings don't differ between components of Velociraptor. This felt like a good idea at that time but may turn out to not be beneficial. Please let me know whether or not I should revert this and leave the API comms untouched.

### CODE REVIEW

File: config/proto/config.proto
Comment (scudette) [inline config/proto/config.proto:262]: It feels like these should go in a CryptoConfig section

File: vql/networking/tls.go
Comment (scudette) [inline vql/networking/tls.go:95]: Maybe uplift this condition so we dont bother with cert hashes unless we need force pinning?

File: vql/networking/tls.go
Comment (scudette) [inline vql/networking/tls.go:78]: Should we fractor this into a smaller function?

**Review by scudette — CHANGES_REQUESTED** [review]
This is very cool. Thanks for sending it. Some minor suggestions only.

File: config/proto/config.proto
Comment (scudette) [inline config/proto/config.proto:261]: Maybe also add the following comment:

To obtain the thumbprint of any server for pinning you can use openssl
openssl s_client -connect www.google.com:443  | openssl x509 -fingerprint -sha1

sha1 Fingerprint=00:94:06:0F:B6:4D:E7:50:72:1E:56:17:9B:C5:85:CF:61:78:42:F1

Also maybe support checking the fingerprint in the format openssl provides to make it more ergonomic for users to just copy/paste the output from openssl into the config file and it should work

File: vql/networking/tls.go
Comment (lucebac) [reply vql/networking/tls.go:95]: I believe this depends on whether or not the idea of being able to short-circuit cert validation may stay (think of the short-circuiting as a means of "if you see this hash, just accept the certificate"). If we did the ForceCertificatePinning check beforehand, short-circuiting wouldn't work anymore because we would skip the hashing when the value is true.

File: vql/networking/tls.go
Comment (lucebac) [reply vql/networking/tls.go:78]: Sounds more than reasonable. I'll see to that on monday.

**Comment by lucebac** [comment]
Seeing my code now makes me think of making the strict cert pinning the default over the less strict short-circuiting mode. I suspect that making strict cert pinning an opt-out makes more sense than it being an opt-in feature. Would you consider that reasonable?

**Comment by scudette** [comment]
what do you mean by less strict short-circuiting vs cert pinning? 

I guess it depends on what the intention of the user is in specifying the conditions in the config file:

If the user provided certificate hashes (btw it might be clearer to call them thumbprint to be more aligned with the openssl terminology) then they are probably saying - i want you to trust all these certs regardless of who signed them. 

But does this mean we never trust any other certs? I think that is a different question because http_client() can be called on anything for multiple reasons (for example downloading third party tools from the client).

So I guess this is my question - does this feature make it more permissive (i.e. allow certs not normally trusted by usual PKI) or less permission - i.e. require only certs in the pinned list to be communicated with. The later case would cause a loss of important functionality - for example we won't be able to serve third party tools from s3 buckets because they can only be downloaded from the Velociraptor server if it is pinned, so this use case needs careful thinking to see if the potential loss of functionality is worth it.

**Comment by lucebac** [comment]
Basically, with the changes in this MR there would be three modes:

* strict mode (certificate pinning): only accept TLS connections if the thumbprint of the certificate is present in the hash list; deny all other connection attempts
* non-strict mode (short-circuiting): bypass certificate validation if the cert has a known thumbprint which is present in the hash list; if the hash is unknown, fall back to the default mode (validation against CaCertificate and public CA lists)
* default mode: certificate validation as it currently is without any modifications

The default mode is chosen when no certificate hashes are present in the config. non-strict mode applies when ForceCertificateValidation is set to false, and strict mode applies when ForceCertificateValidation is set to true.

**Comment by scudette** [comment]
Ok so maybe we need an enum instead of force_certificate_pinning and put it here https://github.com/Velocidex/velociraptor/blob/6c90df816e2d6354f43e9f7686edb5ce513d0865/config/proto/config.proto#L963 named verification_mode or something like that.

The default can be PKI, and the other modes can be PKI_OR_THUMBPRINT and STRICT_THUMBPRINT_PIN or something like that. This makes it clearer which check takes precedence in which scenario. 

We can then refactor the verification mode to the three modes into potentially three helper functions so it is clear what is being done in which scenario. That way the code follows the config settings. It is better to have explicit options in the config rather than imply behavior based on other parameters. 

We can then add a documentation to https://github.com/Velocidex/velociraptor/blob/master/docs/references/server.config.yaml which is presented on the website with extensive descriptions on each selection, including the relevant tradeoffs. Imposing strict pinning essentially disables all http_client() connections to other servers (does this maybe also imply blocking non tls connections too? It might make sense). It is up to the user to decide if the loss of functionality is worth it for them so it is important to adequately describe the tradeoffs in the docs.

**Comment by lucebac** [comment]
I changed the terminology and switch from implicit behavior to using an enum for the verification mode. May I ask you to have a look at the changes?

On a side note: I wasn't able to refactor the thumbprint verification into a dedicated function because of how the return statements work in the verification callback. I had to use a tuple `(bool, error)` as the return value so that I could indicate whether or not we should abort the certificate validation early (the `bool`) and if we want to accept the certificate because, for example, a valid thumbprint was found (`err == nil`) or whether there was something wrong (`err != nil`). This felt to me to be much more complex than it needed to be and so I reverted my changes.

**Comment by scudette** [comment]
Since you did  not enable the Allow edits from maintainers, I pushed my edits to a different branch (you can merge them from there) https://github.com/Velocidex/velociraptor/tree/tls-cert-pinning2

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

The next thing to discuss is if the intention is to lock down access to any VQL accessing a http server when certs are pinned? I noticed you removed the SkipVerify option from the webdav plugin - is the intention to stop people from uploading to a non-pinned server?

**Comment by lucebac** [comment]
Sorry, I always forget to tick that box when creating a MR. I cherry-picked your commits into the branch, they should be visible now. Thanks for helping me out with the refactoring, your changes look great!

The intention is to "force" people to not use unverified connections but instead provide the certificate thumbprint to make sure the connection can be trusted. Since there is the mode `PKI_OR_THUMBPRINT` now, this wouldn't break anything internally but allows to specify these hashes on top of the regular pki validation. It may even be worth to automatically upgrade the default from PKI to PKI_OR_THUMBPRINT once there are thumbprints in the config but I haven't thought through this idea yet.

And yes, locking down all TLS streams to adhere to the verification mode was my ultimate goal. I'm happy to work through the code base and adapt the changes to all places where HTTP(S) is used once the core components of the implementation are completed. For now, I settled with a mwe to keep the changes concise but this could be done as the last step right before a merge.

File: vql/networking/tls.go
Comment (lucebac) [reply vql/networking/tls.go]: Obseleted by using an enum-style approach to the verification mode now.

**Comment by scudette** [comment]
This is the question - and I think it is a good question to discuss. There are lots of places where we allow users of plugins to set SkipVerify so if we want this to be a policy restriction we need to go through them and enforce that the user can not override verification in certain modes.

The practical problem with this is that once clients are deployed then it becomes impossible to use ad-hoc http connections without a config change/redeploy. Similarly on the server it becomes impossible to automatically download third party tool (which is actually a fine hardening step).

**Comment by lucebac** [comment]
You have a point there. Redeploying binaries or changing configs on demand might not be feasible in certain situations. 

Maybe the way to go is somewhere in between: if a user opts for thumbprint-only mode, I believe it's fine to deny SkipVerify since they explicitly went for certificate pinning; for the other two modes (pki with or without thumbprints in addition; be it automatically upgraded or not) it may still be ok to allow users to SkipVerify. I suspect the most restrictive mode is anyway only suitable for users who do very well know what they want to achieve/what they're doing and thus can oversee the implications of their decisions.

**Comment by scudette** [comment]
That is what I was thinking around the EnableSkipVerify() function I introduced. We do need to consistently change all uses of SkipVerify to go through that function. Maybe also we need to return an error in the case where it is denied by policy so users understand why it does not apply the SkipVerify in their particular case.

**Comment by lucebac** [comment]
Yes, we need to let the user know why a connection failed. Especially if we know that it happened because of some config values the user set. I'll work on these two things during the course of the day

**Comment by lucebac** [comment]
I changed all occurences of http.Client, http.Transport, and tls.Config in the vql implementation to use the facilities introduced by this MR. If there's anything I missed, please let me know. I also took the freedom to deprecate all non-SkipVerify flags which disable certificate validation in favor of `SkipVerify` so that the flag has the same name across all plugins. Please let me know if that's ok with you.

**Review by scudette — APPROVED** [review]

**Comment by scudette** [comment]
This looks really good! Thanks !

**Pull request merged**
